### ISSUE TITLE
OpenRTB Native

### ISSUE THREAD

**Issue by dbemiller** [mxmCherry/openrtb#17]
Hey,

Do you plan to support [OpenRTB Native](https://www.iab.com/wp-content/uploads/2016/03/OpenRTB-Native-Ads-Specification-1-1_2016.pdf) objects in this project as well?

If not, we'll probably start up a separate project for it... but if you're interested, we could submit PRs here to help make it happen instead.

**Comment by mxmCherry** [comment]
Yeah, sure, just pushed [feature/native](https://github.com/mxmCherry/openrtb/tree/feature/native) branch. Started it a few weeks ago, just did not have enough "inspiration" to move it forward.

It's still work-in-progress, still a bit messy and, probably, will change much. Also I'll need to double/triple check everything when done (types, comments and so on).

You should definitely check this out first: [github.com/bsm/openrtb/native](https://github.com/bsm/openrtb/tree/master/native)

Very well-written native implementation, IMO. And prod-tested.

I'll do native in this repo as well (and it will be, probably, very much like BSM's implementation, just naming would differ), just don't know when (over a few weekends?..)

**Comment by mxmCherry** [comment]
Just realized it's not very clear from my previous comment - yes, I'd be grateful for native PR(s) :)

Probably, against that `feature/native` branch. Just better keep everything simple (as everything in this repo, roughly - only data/types, no validations (yet), don't think about naming much - copy/paste/camelize stuff from spec), cosmetics can be done later (detailed, nicely-formatted comments etc.)

**Comment by dbemiller** [comment]
sounds good... thanks!

I guess as a followup question, I should ask... what's the motivation for having two different libraries for the same thing? Any philosophical differences which separate the two?

**Comment by mxmCherry** [comment]
Well, I started this repo before I knew about BSM's one (don't think it was private at that time; probably, I just did not try hard enough to find it).

I thought about deleting this one. But it had some stars and traffic - probably, someone really used it.

Also thought about deprecating it. But current repo seemed more accurate to me (in terms of code/documentation/file-organization).

Right now, this repo differs from BSM's one mostly by code-org approaches:
- this repo - quite strict and formalized "think less" approach (naming, files), extensive doc (so usually you don't even have to refer to specs PDF), almost no logic (just types)
- BSM's one - humanize everything (naming, group similar entities into one file - [inventory.go](https://github.com/bsm/openrtb/blob/master/inventory.go) etc), short and laconic documentation, a bit of logic (basic validations and even normalizations)

I think, this repo is closer to something "autogenerated", tbh. Probably, if [Google OpenRTB proto](https://github.com/google/openrtb/blob/master/openrtb-core/src/main/protobuf/openrtb.proto) had better comments (not totally-Google-flavored ones), was written in proto3 (always prefer newer versions) and if proto3 would allow pointer fields (there are no `optional`-s in v3) - I'd deprecate this one in favor of that. Btw, judging from comments (only) in that proto-file, it may not even follow latest (2.5) OpenRTB spec.

Also repos differ by release "policies":
- this one - changed something - released - bumped major version if there's any chance to break consumer code
- BSM's one - emm, just use master, it shouldn't break or will be fixed quickly

Also, this repo seems to be already changed to fit (some / most critical) SSP needs (talking mostly about pointer fields). BSM's one - probably, should fit as well, or can be easily PR-ed; but definitely fits DSP needs.

But let be honest - the main reason to keep this one is mostly because it's my most-starred repo :D
